userspace/falco/grpc_server.cpp

```diff
@@ -0,0 +1,226 @@
+
+		// When event has not been read successfully
+		if(!event_read_success)
+		{
+			if(ctx->m_state != request_context_base::REQUEST)
+			{
+				// todo > log "server completion queue failed to read event for tag `tag`"
+				// End the context with error
+				ctx->end(this, true);
+			}
+			continue;
+		}
+
+		// Process the event
+		switch(ctx->m_state)
+		{
+		case request_context_base::REQUEST:
+			// Completion of m_request_func
+		case request_context_base::WRITE:
+			// Completion of ServerAsyncWriter::Write()
+			ctx->process(this);
+			break;
+		case request_context_base::FINISH:
+			// Completion of ServerAsyncWriter::Finish()
+			ctx->end(this, false);
+			break;
+		default:
+			// todo > log "unkown completion queue event"
+			// todo > abort?
+			break;
+		}
+	}
+}
+
+//
+// Create array of contexts and start processing streaming RPC request.
+//
+#define REGISTER_STREAM(REQ, RESP, RPC, IMPL, CONTEXT_NUM)                          \
+	std::vector<request_stream_context<REQ, RESP>> RPC##_contexts(CONTEXT_NUM); \
+	for(request_stream_context<REQ, RESP> & ctx : RPC##_contexts)               \
+	{                                                                           \
+		ctx.m_process_func = &server::IMPL;                                 \
+		ctx.m_request_func = &output::service::AsyncService::Request##RPC;  \
+		ctx.start(this);                                                    \
+	}
+
+void falco::grpc::server::init(std::string server_addr, int threadiness, std::string private_key, std::string cert_chain, std::string root_certs)
+{
+	m_server_addr = server_addr;
+	m_threadiness = threadiness;
+	m_private_key = private_key;
+	m_cert_chain = cert_chain;
+	m_root_certs = root_certs;
+}
+
+void falco::grpc::server::run()
+{
+	string private_key;
+	string cert_chain;
+	string root_certs;
+
+	falco::utils::read(m_cert_chain, cert_chain);
+	falco::utils::read(m_private_key, private_key);
+	falco::utils::read(m_root_certs, root_certs);
+
+	::grpc::SslServerCredentialsOptions::PemKeyCertPair cert_pair{private_key, cert_chain};
+
+	::grpc::SslServerCredentialsOptions ssl_opts(GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY);
+	ssl_opts.pem_root_certs = root_certs;
+	ssl_opts.pem_key_cert_pairs.push_back(cert_pair);
+
+	::grpc::ServerBuilder builder;
+	builder.AddListeningPort(m_server_addr, ::grpc::SslServerCredentials(ssl_opts));
+	builder.RegisterService(&m_svc);
+
+	m_completion_queue = builder.AddCompletionQueue();
+	m_server = builder.BuildAndStart();
+	falco_logger::log(LOG_INFO, "Starting gRPC server at " + m_server_addr + "\n");
+
+	// The number of contexts is multiple of the number of threads
+	// This defines the number of simultaneous completion queue requests of the same type (service::AsyncService::Request##RPC)
+	// For this approach to be sufficient server::IMPL have to be fast
+	int context_num = m_threadiness * 10;
+	REGISTER_STREAM(output::request, output::response, subscribe, subscribe, context_num)
```

Comment:
If we want to add a second grpc method (like I want to do for coverting k8s psps into falco rules files), how will this work with REGISTER_STREAM and the contexts?

Reply:
It depends on how your proto looks like. You can use `REGISTER_STREAM()` if you have a stream or you will have to make the right macro for different use cases.

Reply:
@mstemm once this is merged we will add a macro to process unary requests as well.
We have in plan to do a top level `version` service to get the falco server version from grpc.

Doesn't make sense for us to add that in this PR, but will be there for when you have to add the psp stuff!